### ISSUE TITLE
Intent to drop the vendor/ folder

### ISSUE THREAD

**Issue by fd0** [restic/restic#2482]
I'd like to drop the `vendor/` folder containing copies of all required libraries used by restic. It was nice to be able to build restic from a checkout of the repo (using `build.go`) without having to fetch external dependencies, here's a list of downsides:
 * It blows up the repository and the source code release
 * It makes reviewing pull requests which add or update dependencies harder because **many** different files are changed
 * It creates merge conflicts for many files (in `vendor`) when master and a PR both change dependencies, in contrast to conflicts in just two files (`go.mod` and `go.sum`)
 
Before Go Modules, vendoring (directly or via a tool like `dep`) was required to manage dependencies. We don't need that any more once we raise the Version required for restic to 1.11. All libraries are fetched for the build process on demand, their integrity is verified by the Go compiler using the hashes contained in `go.mod`, so that's not an issue. For Go >= 1.13, the Go module proxy is automatically activated and we can add instructions the README for the other versions, GitHub repos going away is not a problem any more.

Several other tools have dropped the `vendor/` folder since, and I'm not aware of any incidents or negative user feedback. Did you receive any for Caddy @mholt?

Any thoughts on that?

**Comment by fd0** [comment]
@ncw you have a `vendor/` directory for rclone, what are your thoughts on that?

**Comment by rawtaz** [comment]
I think that while the reasons you outline for having had the vendor/ folder are noble, and I personally don't mind the size of the repo being large (I value it being completely standalone, not depending on anything else but Go), it's fine to make this change if you think it would clean things up.

**Comment by ncw** [comment]
I certainly think the writing is on the wall for the vendor directory!  In go1.13 it doesn't even use the vendor directory unless you set `GO111MODULE=off` or use `go build -mod vendor`.

For the rclone project I like to support 4 go versions which is 2 years of releases.  This lets people on older distros still compile rclone.  So when go1.14 drops, I'll be supporting go1.13, go1.12 and go1.11 which all support modules and then I can drop the vendor directory.

If I was just supporting go1.13 I'd probably drop the vendor directory now as it is actually quite painful to use.  I get contributors to make separate vendor patches so they can be dropped and rebuilt to help with the merge problems. In a long lived PR that sometimes happens several times. And as I said above, go1.13 doesn't touch the vendor directory by default so already your users will be downloading code to satisfy `go.mod` (unless you do something about that in the build process - I didn't check).

To summarise that ramble: if you don't want to suport go1.10 and earlier then I would drop the vendor directory :-)

**Comment by mholt** [comment]
So, Caddy was a bit unique because of its plugin model, and the transition to modules did cause a lot of grief for us because of that. But I think for most repos, it's pretty smooth.

Note that using modules is not mutually exclusive with using the vendor folder.

A couple things to consider, that we've had problems with (not just in Caddy, but at least in any of my projects):

- Even with modules, dependencies can break you by force-pushing or re-tagging. We had this happen once. Naughty of them, but still required an (easy) update to our go.mod and/or our code to stay compatible.

- `go mod vendor` can vendor your modules. Maintaining this, as @ncw mentioned, can be a bit of a bother.

- Vendoring does not provide you any "supply chain" safety guarantees that are superior to Go modules. The only benefit I can think of to using vendoring and modules together is to guarantee that you _can_ compile if dependency repos disappear or violate convention (i.e. force push or re-tag). But modules alone will _prevent_ compilation if your checksums mismatch.

- Having a vendor folder _did_ cause bugs for us at times, due to how Go changes its identifier resolution logic in the presence of a vendor folder.

I personally am not a fan of vendor as a maintainer, but as a user I guess I don't care as long as it works and is well-documented (I actually don't remember all the time how to use go.mod with a vendor folder...)

**Comment by fd0** [comment]
Thanks for the feedback!

> I certainly think the writing is on the wall for the vendor directory! In go1.13 it doesn't even use the vendor directory unless you set GO111MODULE=off or use go build -mod vendor.

Ah, that'll change for 1.14, it will use the `vendor/` automatically if it is there: https://github.com/golang/go/issues/33848

So, overall I think we should drop the vendor dir. Let's see what users say ;)

**Comment by mlusetti** [comment]
I'm far far away to be a go expert but for what is worth I would go with the cleanest and easiest path for the project maintainers/developers.
Reading the comments from others maintainers it seems the smoothest path is to remove the vendor dir so +1 from me.

**Comment by thomasf** [comment]
You could still make `go mod vendor` be a part of the source packaging procedure so that builds from source archives at least still are self contained and can easily be compiled when disconnected from the internet if that is something someone wants.

**Comment by fd0** [comment]
Good point, but before we do that (which makes the build and release process slightly more complicated) I need to know who actually uses that. So let's wait for user feedback. We can always revisit this decision later.

**Comment by fgma** [comment]
I really like the vendoring approach because everything is self-contained. When reading about removing vendor/ this came to my mind: https://www.sciencealert.com/how-a-programmer-almost-broke-the-internet-by-deleting-11-lines-of-code

In terms of security it might not be the best solution to keep local copies of unpatched code but this is more an issue of a package maintainer than the developers.

**Comment by fd0** [comment]
The Go universe has recently introduced a so-called "Module proxy", the most prominent instance is running at https://proxy.golang.org. It caches dependencies to achieve two goals:
 * During build, Go doesn't have to check out the code from each version control system
 * If the upstream goes away, the package stays in the Model proxy

It feels to me that Go has learned from npm and the leftpad disaster :)

**Comment by thomasf** [comment]
Yeah, the only insecurities probably are google having issues or user not being able to access google services (china?) or google deciding to shut down their go mod proxy completely.

**Comment by rumpelsepp** [comment]
> Yeah, the only two main insecurities are probably

Both go away if the source tarball is bundled with the dependencies in /vendor (a bot can upload this tarball for example for each release). The git repo IMO can stay without /vendor.

**Comment by fd0** [comment]
Since I've just looked it up, there are at least two different Module proxys independent of Google:
 * https://goproxy.io/ (apparently hosted by Alibaba)
 * https://goproxy.cn

Those can be configured via `GOPROXY`, e.g. `export GOPROXY=https://goproxy.io,https://goproxy.cn,https://proxy.golang.org,direct`.

When downloading code through these proxys, the Go compiler verifies the integrity of the code. Pretty cool :)
